src/librustc/hir/lowering.rs

```diff
@@ -1479,14 +1479,14 @@ impl<'a> LoweringContext<'a> {
                 let vdata = self.lower_variant_data(vdata);
                 hir::ItemUnion(vdata, self.lower_generics(generics))
             }
-            ItemKind::DefaultImpl(unsafety, ref trait_ref) => {
+            ItemKind::AutoImpl(unsafety, ref trait_ref) => {
```

Comment:
I'd prefer to see all evidence of `impl Trait for ..` go away by the time HIR lowering is done. That is, I think the HIR should not include `impl Trait for ..`-style impls, and instead we should lower any trait that contains such an impl to have `hir::IsAuto` set to true.

This will presumably require one of two things:

- A pre-pass that identifies the set of traits targeted by auto-impls, so that later, when we lower traits, we can set `hir::IsAuto` to true.
- Or, alternatively, we could collect the set as we walk, and update the lowered form of traits after the fact. I suspect we still have unique access and can mutate them in place?